doc/languages-frameworks/javascript.section.md

```diff
@@ -594,6 +594,24 @@ To install the package `yarnInstallHook` uses both `npm` and `yarn` to cleanup p
 
 - `yarnKeepDevDeps`: Disables the removal of devDependencies from `node_modules` before installation.
 
+#### Use with `yarn-berry` and newer lockfiles {#javascript-yarnberry}
+
+The above works well with `yarn` version 1. However, `yarn-berry` introduced a new version of lockfiles, which will not work with classic `yarn`. To still be able to use `fetchYarnDeps` with `yarn-berry`, you will need to supply the source folder containing the `yarn.lock`  and the corresponding project files, and the version of the lockfile:
+
+```nix
+yarnOfflineCache = fetchYarnDeps {
+    src = "${src}/web";
+    yarnVersion = 3;
+    hash = "...";
+  };
```

Comment:
First of all:

```suggestion
yarnOfflineCache = fetchYarnDeps {
  inherit (finalAttrs) src;
  yarnVersion = 3;
  hash = "...";
};
```

Not only that, I noticed that with `yarnVersion`, there should be a `yarnLock` argument to `fetchYarnDeps`, and not `src` argument like here. Do you think you can modify `fetchYarnDeps` a bit such that both usages of the function will get a `yarnLock` argument?

Reply:
It actually needs more than the `lock` file (the `package.json` is mandatory, too, see the test folders)

Reply:
mmh, maybe thats not a bad idea. So we'll use the `yarnLock` argument and derive the directory name from it and use this (internally) as a `src`

Reply:
> mmh, maybe thats not a bad idea. So we'll use the `yarnLock` argument and derive the directory name from it and use this (internally) as a `src`

Exactly that's what I suggest.

Reply:
Ok, I gave it a try, but this is actually harder than I thought. The reason is, that the `yarnLock` file is a _singular_  file in the nixstore, completely independent of the project folder. `yarn-berry` needs both files (it actually wants the project folder, but these two files suffice), but we cannot deduce the project folder from a yarn lockfile which is just a store path. We _need_ a directory (or a separate `packageJson` argument)

Reply:
Small update. I would work with 
```nix
berry_4 = testers.invalidateFetcherByDrvHash fetchYarnDeps {
    yarnLock = "${./berry_4}/yarn.lock";
    yarnVersion = 4;
    sha256 = "sha256-Eq37qKn5P7rkufJBhME0OG/3BsWlCsFYVnWHP5uVhAQ=";
  };
```
This way, the folder get committed to the nix store. I find this version rather more confusing and error prone than using the `src` argument. What do you think?

Reply:
> Ok, I gave it a try, but this is actually harder than I thought. The reason is, that the `yarnLock` file is a _singular_ file in the nixstore, completely independent of the project folder. `yarn-berry` needs both files (it actually wants the project folder, but these two files suffice), but we cannot deduce the project folder from a yarn lockfile which is just a store path. We _need_ a directory (or a separate `packageJson` argument)

Oh I think I have a guess what's going here - you were testing changes to the `berry_{3,4}` attributes at `tests/default.nix`? There, a raw file / directory is currently used (e.g `src = ./berry_3`) - Nix puts that file in the store as is, but in the case of regular packages that have a single `src`, that shouldn't happen. Could you push what you did and I'll try to debug that?

Reply:
sure:
```
diff --git a/pkgs/build-support/node/fetch-yarn-deps/default.nix b/pkgs/build-support/node/fetch-yarn-deps/default.nix
index 90057be587d6..1fe5f1a1906c 100644
--- a/pkgs/build-support/node/fetch-yarn-deps/default.nix
+++ b/pkgs/build-support/node/fetch-yarn-deps/default.nix
@@ -153,7 +153,9 @@ in

               export HOME="$NIX_BUILD_TOP"
               export YARN_ENABLE_TELEMETRY=0
-
+              yarnLock=''${yarnLock:=$PWD/yarn.lock}
+              DIR=$(dirname $yarnLock)
+              cp $DIR/yarn.lock $DIR/package.json $HOME/
               yarn config set enableGlobalCache false
               yarn config set cacheFolder $out
               yarn config set supportedArchitectures --json "$(cat ${./yarn-berry-supported-archs.json})"
diff --git a/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix b/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix
index dda298b51a02..d3864be70ea0 100644
--- a/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix
+++ b/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix
@@ -26,12 +26,12 @@
     sha256 = "sha256-VPnyqN6lePQZGXwR7VhbFnP7/0/LB621RZwT1F+KzVQ=";
   };
   berry_3 = testers.invalidateFetcherByDrvHash fetchYarnDeps {
-    src = ./berry_3;
+    yarnLock = "${./berry_3}/yarn.lock";
     yarnVersion = 3;
     sha256 = "sha256-eiVOpRHuAS7zb6JpCzS3TIHbqQU+wCrMItzZGrE6Fbo=";
   };
   berry_4 = testers.invalidateFetcherByDrvHash fetchYarnDeps {
-    src = ./berry_4;
+    yarnLock = "${./berry_4}/yarn.lock";
     yarnVersion = 4;
     sha256 = "sha256-Eq37qKn5P7rkufJBhME0OG/3BsWlCsFYVnWHP5uVhAQ=";
   };
```

Reply:
> Oh I think I have a guess what's going here - you were testing changes to the `berry_{3,4}` attributes at `tests/default.nix`? There, a raw file / directory is currently used (e.g `src = ./berry_3`) - Nix puts that file in the store as is, but in the case of regular packages that have a single `src`, that shouldn't happen. Could you push what you did and I'll try to debug that?

You're right, this is probably only needed for this artificial test suite and should work fine in a real work scenario. The docs should probably make a note of that. So everyone knows that pointing `yarnLock` to a store path (file) will fail.

Reply:
> This way, the folder get committed to the nix store. I find this version rather more confusing and error prone than using the `src` argument. What do you think?

I too think that using `src` for both yarn classic and yarn berry looks much better, but it'd require changing the function signature for current usages of the function - a somewhat tedious task that I don't feel comfortable asking you to perform. Also, I noticed that `nix build -Lf. fixup-yarn-lock.tests.berry_{3,4}` doesn't work with the diff you just sent.

Reply:
yeah, me too. This works:
```diff
diff --git a/pkgs/build-support/node/fetch-yarn-deps/default.nix b/pkgs/build-support/node/fetch-yarn-deps/default.nix
index 90057be587d6..e6bda1c67038 100644
--- a/pkgs/build-support/node/fetch-yarn-deps/default.nix
+++ b/pkgs/build-support/node/fetch-yarn-deps/default.nix
@@ -153,7 +153,10 @@ in

               export HOME="$NIX_BUILD_TOP"
               export YARN_ENABLE_TELEMETRY=0
-
+              yarnLock=''${yarnLock:=$PWD/yarn.lock}
+              DIR=$(dirname $yarnLock)
+              cp $DIR/yarn.lock $DIR/package.json $HOME/
+              chmod +w $HOME/yarn.lock $HOME/package.json
               yarn config set enableGlobalCache false
               yarn config set cacheFolder $out
               yarn config set supportedArchitectures --json "$(cat ${./yarn-berry-supported-archs.json})"
diff --git a/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix b/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix
index dda298b51a02..d3864be70ea0 100644
--- a/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix
+++ b/pkgs/build-support/node/fetch-yarn-deps/tests/default.nix
@@ -26,12 +26,12 @@
     sha256 = "sha256-VPnyqN6lePQZGXwR7VhbFnP7/0/LB621RZwT1F+KzVQ=";
   };
   berry_3 = testers.invalidateFetcherByDrvHash fetchYarnDeps {
-    src = ./berry_3;
+    yarnLock = "${./berry_3}/yarn.lock";
     yarnVersion = 3;
     sha256 = "sha256-eiVOpRHuAS7zb6JpCzS3TIHbqQU+wCrMItzZGrE6Fbo=";
   };
   berry_4 = testers.invalidateFetcherByDrvHash fetchYarnDeps {
-    src = ./berry_4;
+    yarnLock = "${./berry_4}/yarn.lock";
     yarnVersion = 4;
     sha256 = "sha256-Eq37qKn5P7rkufJBhME0OG/3BsWlCsFYVnWHP5uVhAQ=";
   };
```

Reply:
OK great I'm glad you cracked it. I'll leave to you the decision of whether to change the `fetchYarnDeps` signature now or later. An intermediate option is to support both signatures at the moment, and create a tracking issue that will recommend all maintainers to manually switch to the new signature. When all Nixpkgs usages have transitioned, we can throw an error if `yarnLock` argument is used, to spread the change in usages outside Nixpkgs.

Reply:
Ready for another look :+1: 
Right now, the signature is the same as the existing one (so it will work with a supplied `yarnLock` file), but also accepts a `src` argument. 
After this PR gets merged, I can start a tracking issue to recommend users to switch to `src` for all preexisting `fetchYarnDeps` instances

Reply:
> Right now, the signature is the same as the existing one (so it will work with a supplied `yarnLock` file), but also accepts a `src` argument.
> After this PR gets merged, I can start a tracking issue to recommend users to switch to `src` for all preexisting `fetchYarnDeps` instances

That sounds great, but could you point me please to where the desired `src` signature is treated? I'd like to see if there is a TODO there.

Reply:
I added a TODO above the function signature

Reply:
> I added a TODO above the function signature

OK this is good enough in the meantime.